Added optimizer adam mini

Ashwanth369 opened the pull request:
Implements Adam Mini algorithm as introduced in [Adam-mini: Use Fewer Learning Rates To Gain More] (https://arxiv.org/abs/2406.16793).

Inspired from https://github.com/zyushun/Adam-mini

Ashwanth369 committed:
Added optimizer adam mini

Implements Adam Mini algorithm as introduced in [Adam-mini: Use Fewer Learning Rates To Gain More] (https://arxiv.org/abs/2406.16793).

Datta0 committed:
Fixup ruff linter

Signed-off-by: Dattu Sharma <venkatadattasainimmaturi@gmail.com>

amyeroberts commented:
cc @muellerzr @SunMarc

Ashwanth369 commented:
cc @muellerzr  @SunMarc

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31933). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

muellerzr commented on src/transformers/optimization.py:
```diff
@@ -940,6 +943,325 @@ def get_lr(self):
+        else:
+            self.n_kv_head = self.n_head
+        self.model = model
+        self.world_size = torch.cuda.device_count()
```
Since all distributed stuff runs via `accelerate` to be device-agnostic, we should check here with it instead. 
```suggestion
        self.world_size = PartialState().num_processes
```

muellerzr replied:
(Note the import for earlier is:
```python
if is_accelerate_available():
    from accelerate import PartialState
```

muellerzr commented on src/transformers/optimization.py:
```diff
@@ -940,6 +943,325 @@ def get_lr(self):
+                        if (self.world_size > 1) and (self.model_sharding is True):
+                            tensor_list = [torch.zeros_like(dimension) for _ in range(self.world_size)]
+
+                            dist.all_gather(tensor_list, dimension)
```
Similarly this should use:
```python
from accelerate.utils import gather
```

muellerzr commented on src/transformers/optimization.py:
```diff
@@ -940,6 +943,325 @@ def get_lr(self):
+                    if state["reduced"]:
+                        if "device_mesh" in dir(tmp_lr):
+                            lr_local = tmp_lr.to_local()
+                            dist.all_reduce(lr_local, op=dist.ReduceOp.SUM)
```
This should use
```python
from accelerate.utils import reduce
```

muellerzr commented on src/transformers/optimization.py:
```diff
@@ -940,6 +943,325 @@ def get_lr(self):
+                            dist.all_reduce(lr_local, op=dist.ReduceOp.SUM)
+                            tmp_lr.redistribute(placements=[Replicate()])
+                        else:
+                            dist.all_reduce(tmp_lr, op=dist.ReduceOp.SUM)
```
Similarly

Ashwanth369 replied:
The aforementioned changes have been completed. Kindly review them.

muellerzr reviewed (commented):
Thanks! We should be careful to use accelerate, since the `Trainer` supports a wide-range of backends, not just CUDA. I'm unsure about `Replicate`, so we can leave that be for now and keep an eye if users find that other backends don't support it.

Ashwanth369 committed:
Update optimization.py

Replaced torch.distributed with accelerate for distributed systems.

Update optimization.py

Ashwanth369 committed:
Merge branch 'huggingface:main' into adam-mini

Ashwanth369 committed:
Updated optimization.py

Switched from torch.distributed to accelerate library for distributed operations

Ashwanth369 commented:
@muellerzr

Ashwanth369 commented:
@muellerzr

Ashwanth369 commented:
@muellerzr could you review the changes

Ashwanth369 commented:
@muellerzr

muellerzr reviewed (approved):
Thanks.

amyeroberts commented on src/transformers/optimization.py:
```diff
@@ -940,6 +946,325 @@ def get_lr(self):
         return lrs
 
 
+class AdamMini(Optimizer):
```
There should be a test added in test_optimization.py for this class

amyeroberts commented on src/transformers/optimization.py:
```diff
@@ -940,6 +946,325 @@ def get_lr(self):
+            )
+        if count_output == 0:
+            warnings.warn(
+                "No output layer found.  If you are training Transformers (without weight-tying), please check the name"
```
nit

```suggestion
                "No output layer found. If you are training Transformers (without weight-tying), please check the name"
```

amyeroberts commented on src/transformers/optimization.py:
```diff
@@ -940,6 +946,325 @@ def get_lr(self):
+                " of your output layer and manually add them to 'self.embd_blocks' of Adam-mini. Please ignore this"
+                " warning if you are using weight-tying."
+            )
+        if count_q == 0:
+            warnings.warn(
+                "No Query found.  If you are training Transformers, please check the name of your Query in attention"
+                " blocks and manually add them to 'self.qk_blocks' of Adam-mini"
+            )
+
+        if count_k == 0:
+            warnings.warn(
+                "No Key found.  If you are training Transformers, please check the name of your Key in attention blocks"
+                " and manually add them to 'self.qk_blocks' of Adam-mini"
+            )
```
```suggestion
        if count_q == 0:
            warnings.warn(
                "No Query found. If you are training Transformers, please check the name of your Query in attention"
                " blocks and manually add them to 'self.qk_blocks' of Adam-mini"
            )

        if count_k == 0:
            warnings.warn(
                "No Key found. If you are training Transformers, please check the name of your Key in attention blocks"
                " and manually add them to 'self.qk_blocks' of Adam-mini"
            )
```

amyeroberts reviewed (commented):
Thanks for adding! 

Only thing is the need for a test

Ashwanth369 committed:
Merge branch 'huggingface:main' into adam-mini

Ashwanth369 committed:
Added test case for Adam Mini

Ashwanth369 committed:
Fixed repository consistency errors

ValueError: The main __init__ has objects that are not present in transformers.utils.dummy_pt_objects.py. Run `make fix-copies` to fix this.

Ashwanth369 committed:
Documentation for AdamMini

Ashwanth369 commented:
@amyeroberts completed the mentioned changes. Could you review them

amyeroberts commented on tests/optimization/test_optimization.py:
```diff
@@ -110,6 +111,49 @@ def test_adafactor(self):
+        with torch.no_grad():
+            preds = model(inputs)
+            loss = criterion(preds.squeeze(), targets)
+        self.assertListAlmostEqual([loss], [0.0], tol=0.01)
```
Is this right? A loss of 0 seems unlikely. What is the loss after the first step?

Ashwanth369 replied:
Typically, the loss begins around 0.4-0.5 and eventually drops below 0.01. That's why I used that value as the tolerance.

amyeroberts reviewed (commented):
Thanks for updating and adding a test! Just a question on the test values

Ashwanth369 commented:
@amyeroberts

amyeroberts commented on src/transformers/optimization.py:
```diff
@@ -940,6 +946,325 @@ def get_lr(self):
+
+    def __init__(
+        self,
+        model: Union[PreTrainedModel, nn.Module],
```
Why does this optimizer take `model` here instead of `params` like the other optimizers?

Ashwanth369 replied:
Adam Mini uses a single momentum and variance for groups of parameters, rather than for each one individually like Adam does. These groups are usually based on layers or modules. To make this work, Adam Mini takes the model and creates these shared parameters for each layer or module instead of doing it for every single parameter.

amyeroberts replied:
I think to be able to use this with the library the class will need to have the same API as other optimizers. Could you run a test with trainer to see if it integrates properly:

```py
config = LlamaConfig(
    vocab_size=100, 
    hidden_size=32, 
    num_hidden_layers=3, 
    num_attention_heads=4
)
tiny_llama = LlamaForCausalLM(config)
x = torch.randint(0, 100, (128,))
train_dataset = RepeatDataset(x)

with tempfile.TemporaryDirectory() as tmpdir:
    # Trainer without inf/nan filter
    args = TrainingArguments(
        tmpdir,
        learning_rate=1e-9,
        logging_steps=5,
        optim="adam_mini",
    )
    trainer = Trainer(tiny_llama, args, train_dataset=train_dataset)

    # Check this works
    _ = trainer.train()
```

Mrw33554432 replied:
> I think to be able to use this with the library the class will need to have the same API as other optimizers. Could you run a test with trainer to see if it integrates properly:
> 
> ```python
> config = LlamaConfig(
>     vocab_size=100, 
>     hidden_size=32, 
>     num_hidden_layers=3, 
>     num_attention_heads=4
> )
> tiny_llama = LlamaForCausalLM(config)
> x = torch.randint(0, 100, (128,))
> train_dataset = RepeatDataset(x)
> 
> with tempfile.TemporaryDirectory() as tmpdir:
>     # Trainer without inf/nan filter
>     args = TrainingArguments(
>         tmpdir,
>         learning_rate=1e-9,
>         logging_steps=5,
>         optim="adam_mini",
>     )
>     trainer = Trainer(tiny_llama, args, train_dataset=train_dataset)
> 
>     # Check this works
>     _ = trainer.train()
> ```

I think they at least need to get a `named_parameters`, like
```
from adam_mini import Adam_mini

optimizer = Adam_mini(
            named_parameters = model.named_parameters(), 
            lr = lr, 
            betas = (beta1,beta2), 
            eps = eps,
            weight_decay = weight_decay, 
            model_sharding = True,
            dim = model_config.dim,
            n_heads = model_config.n_heads,
            n_kv_heads = model_config.n_kv_heads,
            )
```
I think it's impossible to get what they need from the parameters passed to the optimizer. And as long as there is not an optimizer augments, any type of workarounds also seems unfeasible.

Ashwanth369 replied:
I’m sorry, but I’m having trouble understanding your point. There are other optimizers, like Lomo and AdaLomo, that also use a model as input instead of parameters. I’m unsure if the issue is due to the different API formats of the optimizers in optimization.py, which might cause Adam Mini to be inconsistent, or if the real problem is that we shouldn’t be passing the model as a parameter to the optimizer. Could you clarify this for me?

Mrw33554432 replied:
> I’m sorry, but I’m having trouble understanding your point. There are other optimizers, like Lomo and AdaLomo, that also use a model as input instead of parameters. I’m unsure if the issue is due to the different API formats of the optimizers in optimization.py, which might cause Adam Mini to be inconsistent, or if the real problem is that we shouldn’t be passing the model as a parameter to the optimizer. Could you clarify this for me?

I guess her point is that all current optimizers in transformers lib pass model.parameters as input, and she want Adam mini to stick with that.

amyeroberts replied:
@Mrw33554432 Right, or at least there should be a test which confirms this optimizer can be selected and run correctly in `Trainer` with `optim="adam_mini"` set

Ashwanth369 replied:
@amyeroberts  Instead of passing the model as an argument, can I pass the named parameters as @Mrw33554432 suggested? If that's not possible, I can go ahead and add the test case with Trainer as you proposed.

amyeroberts replied:
I'd actually like you to do both: pass in the parameters and check that the optimizer can be used in trainer as expected. For the second case, you don't have to add a test, but we should confirm it works. The test that's added here should be updated to have the same structure as the other tests for custom optimizers i.e. the same structure as `test_adafactor`

Ashwanth369 replied:
I've written the test case for the trainer, and it's working as expected. However, I can't pass named parameters directly to the optimizer and instead have to pass the entire model. Is there any other workaround for this?

Ashwanth369 committed:
Updated test case for Adam Mini

Ashwanth369 committed:
Fixup ruff linter

Fixup ruff linter

Ashwanth369 commented:
@amyeroberts could you check the new test case and let me know if that works?

liziniu commented:
Hi, all. Thanks for implementing Adam-mini into Transformers!

I am one of the authors of the Adam-mini paper. Adam-mini is now available on PyPI (i.e., pip install adam_mini). We may update the optimizer in the future, so we prefer to support AdamW in Transformers by importing Adam-mini directly. We already have such an implementation. Could you provide some tips for us? Should I create a new pull request?

liziniu commented:
Hi all,

I have discussed this with my collaborators, and we have a proposal. We suggest naming the optimizer in this pull request 'Adam_mini_HF' because we think it is really cool to make it compatible with `accelerate`. At the same time, I will create a new pull request that supports using Adam-mini directly via importing Adam-mini directly.

My code for this version of Adam-mini is available [here](https://github.com/liziniu/transformers/tree/adam_mini) and is ready for a new pull request.

Key code sections are highlighted below:

https://github.com/liziniu/transformers/blob/adam_mini/src/transformers/trainer.py#L1125-L1129

https://github.com/liziniu/transformers/blob/adam_mini/src/transformers/trainer.py#L1221-L1244

Best,
Ziniu

Ashwanth369 commented:
> Hi all,
> 
> I have discussed this with my collaborators, and we have a proposal. We suggest naming the optimizer in this pull request 'Adam_mini_HF' because we think it is really cool to make it compatible with `accelerate`. At the same time, I will create a new pull request that supports using Adam-mini directly via importing Adam-mini directly.
> 
> My code for this version of Adam-mini is available [here](https://github.com/liziniu/transformers/tree/adam_mini) and is ready for a new pull request.
> 
> Key code sections are highlighted below:
> 
> https://github.com/liziniu/transformers/blob/adam_mini/src/transformers/trainer.py#L1125-L1129
> 
> https://github.com/liziniu/transformers/blob/adam_mini/src/transformers/trainer.py#L1221-L1244
> 
> Best, Ziniu

Ashwanth369 commented:
@liziniu sounds great! I will go ahead and change the name to "Adam_mini_HF". @amyeroberts does this work?

amyeroberts commented:
@Ashwanth369 @liziniu in this case, I think the more canonical integration would be to use the optimizer from the source library and just use like e.g. [like for GrokAdamW](https://github.com/huggingface/transformers/blob/47b096412da9cbeb9351806e9f0eb70a693b2859/src/transformers/trainer.py#L1458). I don't have a strong opinion here - maybe @muellerzr can weigh in on what the best balance is in terms of functionality v.s. simplicity
